File: packages/sdk-typescript/test/webhooks.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { createHmac } from "node:crypto";
import { verifyWebhookSignature } from "../dist/webhooks.js";

const SECRET = "whsec_test";

// Mirror of the gateway's signing (internal/webhooks signing.go):
// HMAC-SHA256 over `${timestamp}.${nonce}.${body}`, base64url, "v1=" prefix.
function sign(timestamp, nonce, body) {
  const base = `${timestamp}.${nonce}.${body}`;
  const digest = createHmac("sha256", SECRET).update(base).digest("base64url");
  return `v1=${digest}`;
}

test("verifyWebhookSignature accepts a valid recent gateway signature", async () => {
  const ts = Math.floor(Date.now() / 1000).toString();
  const body = '{"event":"job.completed"}';
  const sig = sign(ts, "nonce_01", body);
  assert.equal(
    await verifyWebhookSignature(Buffer.from(body), sig, SECRET, { timestamp: ts, nonce: "nonce_01" }),
    true,
  );
});

test("verifyWebhookSignature rejects a wrong nonce (replayed signature)", async () => {
  const ts = Math.floor(Date.now() / 1000).toString();
  const body = '{"event":"job.completed"}';
  const sig = sign(ts, "nonce_01", body);
  assert.equal(
    await verifyWebhookSignature(Buffer.from(body), sig, SECRET, { timestamp: ts, nonce: "nonce_02" }),
    false,
  );
});

test("verifyWebhookSignature rejects a wrong signature", async () => {
  const ts = Math.floor(Date.now() / 1000).toString();
  assert.equal(
    await verifyWebhookSignature(Buffer.from("x"), "v1=deadbeef", SECRET, { timestamp: ts, nonce: "n" }),
    false,
  );
});

test("verifyWebhookSignature rejects the version-less hex format", async () => {
  const ts = Math.floor(Date.now() / 1000).toString();
  const body = "x";
  // Old SDK bug: hex over `${timestamp}.${body}` — must no longer verify.
  const legacy = createHmac("sha256", SECRET).update(`${ts}.${body}`).digest("hex");
  assert.equal(
    await verifyWebhookSignature(Buffer.from(body), legacy, SECRET, { timestamp: ts, nonce: "n" }),
    false,
  );
});

test("verifyWebhookSignature rejects an expired timestamp", async () => {
  const ts = (Math.floor(Date.now() / 1000) - 10 * 60).toString(); // 10 min old
  const body = "x";
  const sig = sign(ts, "nonce_01", body);
  assert.equal(
    await verifyWebhookSignature(Buffer.from(body), sig, SECRET, { timestamp: ts, nonce: "nonce_01" }),
    false,
  );
});

File: packages/sdk-go/webhooks_test.go
package ubag

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"testing"
	"time"
)

// signGatewayStyle mirrors the gateway's signing (internal/webhooks
// signing.go): HMAC-SHA256 over `${timestamp}.${nonce}.${body}`, base64url,
// "v1=" prefix.
func sign(secret, ts, nonce, body string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(fmt.Sprintf("%s.%s.%s", ts, nonce, body)))
	return "v1=" + base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}

func TestVerifyWebhookValid(t *testing.T) {
	ts := fmt.Sprintf("%d", time.Now().Unix())
	body := `{"event":"job.completed"}`
	sig := sign("whsec", ts, "nonce_01", body)
	if !VerifyWebhookSignature([]byte(body), sig, "whsec", ts, "nonce_01", 300) {
		t.Fatal("expected valid gateway signature to verify")
	}
}

func TestVerifyWebhookWrongNonceRejected(t *testing.T) {
	ts := fmt.Sprintf("%d", time.Now().Unix())
	body := "x"
	sig := sign("whsec", ts, "nonce_01", body)
	if VerifyWebhookSignature([]byte(body), sig, "whsec", ts, "nonce_02", 300) {
		t.Fatal("expected wrong nonce (replay) to fail")
	}
}

func TestVerifyWebhookBadSig(t *testing.T) {
	ts := fmt.Sprintf("%d", time.Now().Unix())
	if VerifyWebhookSignature([]byte("x"), "v1=deadbeef", "whsec", ts, "nonce", 300) {
		t.Fatal("expected bad signature to fail")
	}
}

func TestVerifyWebhookLegacyHexRejected(t *testing.T) {
	ts := fmt.Sprintf("%d", time.Now().Unix())
	body := "x"
	// Old SDK bug: hex over `${timestamp}.${body}` — must no longer verify.
	legacyMac := hmac.New(sha256.New, []byte("whsec"))
	legacyMac.Write([]byte(ts + "." + body))
	legacy := fmt.Sprintf("%x", legacyMac.Sum(nil))
	if VerifyWebhookSignature([]byte(body), legacy, "whsec", ts, "nonce", 300) {
		t.Fatal("expected legacy hex signature format to fail")
	}
}

func TestVerifyWebhookExpired(t *testing.T) {
	ts := fmt.Sprintf("%d", time.Now().Add(-10*time.Minute).Unix())
	body := "x"
	sig := sign("whsec", ts, "nonce_01", body)
	if VerifyWebhookSignature([]byte(body), sig, "whsec", ts, "nonce_01", 300) {
		t.Fatal("expected expired timestamp to fail")
	}
}
